ping/pinger.go

```diff
@@ -0,0 +1,189 @@
+			return icmp.ListenPacket(network, address)
+		},
+	}
+
+	p.sndTicker = time.NewTicker(inteval)
+	p.timeout = timeout
+
+	p.addr = net.UDPAddr{
+		IP:   ip.IP,
+		Zone: ip.Zone,
+	}
+	r := rand.New(rand.NewSource(time.Now().UnixNano()))
+	p.seq = r.Intn(1 << 16)
+
+	return &p, nil
+}
+
+// Ping will continuously ping the configured target until the context is cancled
+func (p *pinger) Ping(ctx context.Context) error {
+	l := logr.FromContextOrDiscard(ctx).WithValues("target", p.addr.String())
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	conn, err := p.listen("udp4", "")
+	if err != nil {
+		return fmt.Errorf("Failed to listen for ICMP packets: %w", err)
+	}
+	p.conn = conn
+	defer conn.Close()
+
+	go func() {
+		defer cancel()
+		err := p.receive(ctx)
+		if err != nil {
+			l.Error(err, "Pinger receiver failed")
+		}
+	}()
+
+	for {
+		select {
+		case <-p.sndTicker.C:
+			err := p.ping()
+			if err != nil {
+				l.Error(err, "Pinger failed")
+				return err
+			}
+			p.scrubInflight(ctx)
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+func (p *pinger) scrubInflight(ctx context.Context) {
+	l := logr.FromContextOrDiscard(ctx)
+	p.inflight.Range(func(key, value any) bool {
+		sent, ok := value.(time.Time)
+		if !ok {
+			l.Error(errors.New("inconsistent datastructure"), "Found malformed entry in inflight packet map. Dropping entry.")
+			p.inflight.Delete(key)
+		}
+		if sent.Add(p.timeout).Before(now()) {
+			p.obs.WithLabelValues("lost").Observe(p.timeout.Seconds())
+			p.inflight.Delete(key)
+		}
+		return true
+	})
+}
+
+func (p *pinger) receive(ctx context.Context) error {
+	l := logr.FromContextOrDiscard(ctx)
+	for {
+		res, err := p.recv()
+		if err != nil && ctx.Err() == nil {
+			l.Error(err, "Failed to receive ping response")
+			if netErr, ok := err.(net.Error); !ok || !netErr.Timeout() {
+				return err // socket gone
+			}
+		}
+		if ctx.Err() != nil {
+			return nil
+		}
+
+		if _, ok := p.inflight.LoadAndDelete(res.Seq); ok {
+			received := now()
+			sent := time.UnixMicro(int64(binary.BigEndian.Uint64(res.Data)))
+			l.V(1).Info("got reflection", "id", res.ID, "seq", res.Seq, "sent", sent, "received", received, "latency", received.Sub(sent))
+			p.obs.WithLabelValues("completed").Observe(received.Sub(sent).Seconds())
+			p.inflight.Delete(res.Seq)
+		}
+	}
+}
+
+func (p *pinger) ping() error {
+	sent := now()
+	ts := make([]byte, 8)
+	binary.BigEndian.PutUint64(ts, uint64(sent.UnixMicro()))
+
+	e := icmp.Echo{
+		ID:   0, // Ignored for unpriv ICMP sockets
+		Seq:  p.seq,
+		Data: ts,
+	}
+	msg := icmp.Message{
+		Type: ipv4.ICMPTypeEcho,
+		Code: 0,
+		Body: &e,
+	}
+	msgB, err := msg.Marshal(nil)
+	if err != nil {
+		return fmt.Errorf("Failed to marshal ICMP packet: %w", err)
+	}
+
+	if _, err := p.conn.WriteTo(msgB, &p.addr); err != nil {
+		return fmt.Errorf("Failed to send ICMP packet: %w", err)
+	}
+	p.inflight.Store(p.seq, sent)
+	p.seq = p.seq + 1
```

Comment:
The overflow is a valid question, I suspect that should roll over fine, but you're right we should be explicit.

However I don't really see the reason for the global counter:

* We don't really depend on uniqueness. I really don't yet see the error case. The id is unique per socket that really shouldn't lead to problems
* Also 16bit randomness should be enough. There are only ever max 5 packets in flight per node. The probability of a collision should be low (I'd need to properly calculate that tough)

Reply:
Oke I finally understood the issue, I think (https://github.com/prometheus/blackbox_exporter/issues/411 has a good example).

However I don't see how a global counter solves this. If we would ping a NATed address through a gateway that also NATs ICMP packets wrong, having a global counter per pod doesn't help us really. The problem would be if two canary pods choose the same seq and id, which is possible.

We would need a global counter for all canary instances, and I  don't really want to go down that rabbithole..

Reply:
It really depends on how often targets are added. Every time targets is added there's a chance of overlapping offset. With not much target churn it will be mostly fine. Still, as we are doing stuff that might alert at night I'm on the cautious side here.

The global counter could be just like `time.Now()` and mocked the same.

> We would need a global counter for all canary instances, and I don't really want to go down that rabbithole..

Nope we do not want that 😅 I indeed had the issue with Palo Alto FWs but yeah irrelevant for us.

Reply:
There is one conn per pinger and the kernel manages the multiplexing through the `ID` field. Thus the session ID is not very relevant for us. Nothing will break with overlapping session ids.

Reply:
I found the kernel patch for this: https://lwn.net/Articles/420800/

TL;DR the way we currently use this, the kernel chooses a free ICMP ID similarly to choosing a free UDP port when opening a socket with port 0. So each `conn` has a unique ID and the kernel will multiplex them. (We could also use custom IDs by setting the port in the UDP address I think, but I don't see a use-case for that right now)